core/src/main/scala/shapeless/MacroState.scala

```diff
@@ -0,0 +1,48 @@
+package shapeless
+
+import scala.reflect.{ClassTag, classTag}
+import scala.reflect.macros.Universe
+
+
+private[shapeless] object MacroState {
+  private class MacroStateAttachment(val state: collection.mutable.HashMap[Class[_], Any])
+
+  /**
+   * Associates some state with a global.Run. Preferable to using static state in the macro
+   * classloader (e.g vars in top level objects) so as to avoid race conditions under `-Ycache-macro-classloader`.
+   *
+   * @tparam T the type of the state. The class tag for this type will be used a a map key, so each user of
+   *           this facility should use a dedicated and distinct class to wrap the state.
+   * @param u The reflection universe, typically obtained from `context.universe` in a macro implementation.
+   * @param factory Factory to create the state.
+   */
+  def apply[T: ClassTag](u: Universe, factory: => T): T = {
```

Comment:
```suggestion
  def getOrElseUpdate[T: ClassTag](u: Universe, factory: => T): T = {
```

... except the key isn't the first param, it's `ClassTag[T].runtimeClass`.

But, still, I think it might help explain intent a bit better.